# Week 01 Assignment genes expressed in the brain

Welcome to the preparatory programming course! Here, you'll learn to develop software programs using the basics of Python. You'll explore Python's built-in functions to tackle assignments, giving you a solid grasp of how data manipulation works right from the ground up. Forget about fancy tools like pandas and numpy in this course – we're sticking to the simple yet powerful features of Python.

By diving into these foundational Python concepts, you'll sharpen your skills for solving problems with logical thinking and turning those solutions into functional code. These skills are fundamental building blocks of programming, and you'll be mastering them in the coming 7 weeks. This first week we will look into reading and manipulating files. 

By the end of this week you should be able to translate a simple datadriven question into a working Python program. 

Learning outcomes of this week are:
- can use the Linux shell to navigate and do simple text processing
- use the correct Python data types to store data
- know how to use flow control
- can read a text file into the Python environment
- can use f-string formatting


For this course we will use data that was made available through the Allen Institute for Brain Science. For the assignment of week 1 we are going to have a look at microarray expression data taken from multiple locations (samples) in the brain of one individual. Please have a look at the experimental setup using the following link: [background reading](https://help.brain-map.org/download/attachments/2818165/WholeBrainMicroarray_WhitePaper.pdf?version=1&modificationDate=1433804560874&api=v2)

If you are not familiar with microarray expression data and how it is generated you might want to watch this [background video](https://www.youtube.com/watch?v=xBr2qznEaqw)


The end goal of this week is a program that will read a file containing measured expression values from probes capturing active genes in the human brain. From all active genes in the brain we would like to know:
- which gene has the highest maximum expression over all samples measured
- which gene has the highest average expression over all samples 


> NOTE: All assignments should be done on our Linux systems, this to ensure that you have familiarized yourself with our systems. This prepares you for the rest of the courses of the master program.


The assignment consists of 5 parts:
- [Part 1: Getting the data](#Part-1:-Getting-the-data)
- [Part 2: Reading the data](#Part-2:-Read-the-data)
- [Part 3: Preparing the data](#Part-3:-Prepare-the-data)
- [Part 4: Processing the data](#Part-4:-Process-the-data)
- [Part 5: Integrating other datasources](#Part-5:-Integrate-other-sources-of-information)
- [Part 6: Bonus](#Part-6:-Bonus-(optional))


## Part 1: Getting the data
Please download the zip file `H0351.2001` from the Complete normalized microarray datasets at https://human.brain-map.org/static/download   
Unpack this file and place the content of the unpacked folder inside the `Data` folder of this repository. It should contain the following files

- Readme.txt
- Ontology.csv
- Probes.csv
- PACall.csv
- MicroarrayExpression.csv
- SampleAnnot.csv

## Part 1.1: Inspecting the data using the Linux command line
Using the Linux terminal (top left corner -> Terminal Emulator or task bar at the bottom of your screen 2nd Icon):
- navigate to the location where you placed the unpacked folder
- read the `Readme.txt` from your linux terminal and describe in your own words what the content is of the files:
    - `MicroarrayExpression.csv`
    - `Probes.csv`
- Make sure that you explain what probes and samples are in the context of the research and how the files are related to each other. If you are not familiar with microarray expression data and how it is generated you might want to watch this [video](https://www.youtube.com/watch?v=xBr2qznEaqw)

For each file listed above, show the first 5 lines to the screen. Is it clear from the `Readme.txt` what you are looking at?

> Tip: on our Debian linux systems the shell is `bash` and a quick google search with the term `bash commands` will show you many sites explaining the most used bash commands.

In [1]:
head MicroarrayExpression.csv

In [2]:
# Use this cell to describe the files in words 

## Part 2: Read the data
Write Python code that will read the `MicroarrayExpression.csv` file line by line. Remember: You can **not** use the `pandas` or `csv` library but yoiu have to use the built in Python funtion `with open()`. After that you `print` the *first* line to the screen

> Note: It is bad habbit to completely read a file into memory as this will keep many MB's or even GB's of space occupied. Instead, read a file line by line and process the content as each line is being read.

Expected outcome:

1058685,3.6157916807089787,2.13807367935382,2.4805415023588107.....3.3062090302215705

In [1]:
#open the file read the lines
#code page determined by using 'file -bi <filename>' in bash
with open('MicroarrayExpression.csv', encoding='us-ascii') as ma_expressions:
    #read the first line and store it in the created variable
    first_line = ma_expressions.readline()

print(first_line)


1058685,3.6157916807089787,2.13807367935382,2.4805415023588107,2.96497183239022,2.67980316976351,1.8562381007421844,2.28043513665554,3.08085727123301,2.62857527091796,2.358911706578,3.1134311343748804,2.2332099397537006,2.43376313475427,3.79115692339681,2.7335022860878,2.57679469424082,2.39851667566474,2.28475938335585,4.44402647195994,2.5213198933884597,2.49313157687943,2.39608892908155,2.51249881281416,2.40144577175088,2.04032083382348,2.3078927851154,2.43847554381551,2.19312922079118,3.23146250399306,2.931276590737998,2.39231145706868,2.61591085981625,3.9953495214663004,2.07377883844337,2.82259662852244,3.22470487544826,2.57869657335114,2.66137403742995,2.61710393614862,3.71237805540011,2.31104606348884,2.64800304561617,3.671856592108209,2.3719628094516,2.41077173194952,2.557942539671,2.3941417217615735,4.184898640318601,3.183660214148734,2.7223544675944904,4.04157694433362,4.086520221001682,3.99772546908317,3.83955825852816,4.0985907461571,3.68024602035813,3.865413670799189,4.25988

In [2]:
'''a possible solution for part 2'''
with open('MicroarrayExpression.csv', encoding='us-ascii') as ma_expressions:
    # read the first line and print it
    print(ma_expressions.readline())

1058685,3.6157916807089787,2.13807367935382,2.4805415023588107,2.96497183239022,2.67980316976351,1.8562381007421844,2.28043513665554,3.08085727123301,2.62857527091796,2.358911706578,3.1134311343748804,2.2332099397537006,2.43376313475427,3.79115692339681,2.7335022860878,2.57679469424082,2.39851667566474,2.28475938335585,4.44402647195994,2.5213198933884597,2.49313157687943,2.39608892908155,2.51249881281416,2.40144577175088,2.04032083382348,2.3078927851154,2.43847554381551,2.19312922079118,3.23146250399306,2.931276590737998,2.39231145706868,2.61591085981625,3.9953495214663004,2.07377883844337,2.82259662852244,3.22470487544826,2.57869657335114,2.66137403742995,2.61710393614862,3.71237805540011,2.31104606348884,2.64800304561617,3.671856592108209,2.3719628094516,2.41077173194952,2.557942539671,2.3941417217615735,4.184898640318601,3.183660214148734,2.7223544675944904,4.04157694433362,4.086520221001682,3.99772546908317,3.83955825852816,4.0985907461571,3.68024602035813,3.865413670799189,4.25988

## Part 3: Prepare the data
Each line consists of 2 seperate parts, the first element of each line is the probe identifier and all other elements are measured microaray expression levels. For this part we want to have to **probe IDs** and the **rest** seperated from each other and stored in **two variables**. Please use variables names according the [PEP8 style guide](https://pep8.org/#introduction)

Add the Python code that should be able to do so in the code block below. You are allowed to copy the code from part 2 and modify the code in order to create and fill the two variables. 
Again verify that the output is correct by printing the first line to the screen. Limit the number of expression values printed to screen to 10

Expected outcome (depending on the variable data types you choose) 

1058685 ['3.6157916807089787', '2.13807367935382', '2.4805415023588107', '2.96497183239022', '2.67980316976351', '1.8562381007421844', '2.28043513665554', '3.08085727123301', '2.62857527091796', '2.358911706578']

In [3]:
def split_line(line):
    '''Goal:
        this functions splits teh contents of the input string on the first comma
        and returns a tuple consisting of the part before the comma and a part
        following the comma
       Input: string that will be split
       Output: tuple containing both parts of the split string
       Preconditions: -
       Postconditions: -
    '''
    #search the position of the first comma
    first_comma_pos = line.find(',')

    #get the fist part
    probe_id = line[0:first_comma_pos]

    #get the second part WITHOUT the comma
    expr_values = line[first_comma_pos + 1:]
    #convert second part to a list at the comma positions
    expr_values = expr_values.split(',')

    #remove the superfluous entries
    expr_values = expr_values[:9]

    #for part 3.1. convert to integers using enumerate and not range because of pylint
    for pos, item in enumerate(expr_values):
        expr_values[pos] = float(item)
    #return both parts
    return probe_id, expr_values

#open the file read and process the first line
with open('MicroarrayExpression.csv', encoding='us-ascii') as ma_expressions:
    #read the first line, process it in the previously created function
    #and store the results in variables
    pid, values = split_line(ma_expressions.readline())
#print
print(pid, values[:10])




1058685 [3.6157916807089787, 2.13807367935382, 2.4805415023588107, 2.96497183239022, 2.67980316976351, 1.8562381007421844, 2.28043513665554, 3.08085727123301, 2.62857527091796]


In [11]:
'''alternative solution'''
with open('MicroarrayExpression.csv', encoding='us-ascii') as ma_expressions:
    #read the first line, process it and store it in the created variable
    #split it using the unpack operator and slice it to print only 10 items
    pid, *values = ma_expressions.readline().split(',')[:10]

print(pid, values)


1058685 ['3.6157916807089787', '2.13807367935382', '2.4805415023588107', '2.96497183239022', '2.67980316976351', '1.8562381007421844', '2.28043513665554', '3.08085727123301', '2.62857527091796']


## Part 3.1: More data preperation
To calculate the average and maximum of each probe, we first have to convert each expression to a suitable Python data type.

Implement the code that will do this in the code block below. You are allowed to copy the code from part 3 and modify the code in order to create and fill the two variables with the suitable data type. You can **not** use the libraries pandas or numpy for this. Again just as above print from the first line the probe identifier and the first 10 (converted) measured expression values

Expected outcome (depending on the variable data type you choose): 

1058685 [3.6157916807089787, 2.13807367935382, 2.4805415023588107, 2.96497183239022, 2.67980316976351, 1.8562381007421844, 2.28043513665554, 3.08085727123301, 2.62857527091796, 2.358911706578]

In [7]:
'''alternative solution'''
with open('MicroarrayExpression.csv', encoding='us-ascii') as ma_expressions:
    #read the first line, process it and store it in the created variable
    #split it using the unpack operator and slice it to print only 10 items
    pid, *values = ma_expressions.readline().split(',')[:10]

#for part 3.1. convert to integers using enumerate and not range because of pylint
for pos, item in enumerate(values):
    values[pos] = float(item)
print(pid, values)


probe: 1058685, max expression: 5.266, average expression 3.171


## Part 4: Process the data
For each probe we would like to calculate the maximum value and the average value. Incorporate the Python code that will calculate these two values for every probe.

Print from the first line the probe identifier and the calculated maximum and average expression value. Round the expression values to 3 decimal places. Tip: use fstring formatting.  

Expected outcome (something similar to):

probe: 1058685, max expression: 5.266, average expression: 3.171

In [9]:
#Of course I could have written the code to calculate the mean myself,
#but 'a good programmer is a lazy programmer' so I use what is available
#in the standard library
from statistics import mean

def process_data(line):
    '''processes a line a specified in assignments of week 1, part 4
        input: string which is a line consisting of 1 int (the id) and many floats (the values)
        output: a tuple consisting of the id, maximum value and average value of the values
    '''
    pid, *values = line.split(',')
    for position, value in enumerate(values):
        values[position] = float(value)
    return (pid, round(max(values), 3), round(mean(values), 3))


with open('MicroarrayExpression.csv', encoding='us-ascii') as ma_expressions:
    #gather all results in a list
    result = process_data(ma_expressions.readline())

#print the first line
print(f'probe: {result[0]}, max expression: {result[1]}, average expression {result[2]}')




probe: 1058685, max expression: 5.266, average expression 3.171


## Part 4.1: Report on analysis
Report on (print to screen):
- the probe with the highest maximum expression value measures over all samples and the highest average expression value over all samples.

Expected outcome: 

There are multiple probes with a maximum expression of 18.3817518519472, the probe with the highest average expression is probe 1070089, which has also the maximum expression of 18.3817518519472. The expressions should be printed in 3 decimal

In [8]:
#Of course I could have written the code to calculate the mean myself,
#but 'a good programmer is a lazy programmer' so I use what is available
#in the standard library
from statistics import mean

def process_data(line):
    '''processes a line a specified in assignments of week 1, part 4
        input: string which is a line consisting of 1 int (the id) and many floats (the values)
        output: a tuple consisting of the id, maximum value and average value of the values
    '''
    pid, *values = line.split(',')
    for position, value in enumerate(values):
        values[position] = float(value)
    return (pid, round(max(values), 3), round(mean(values), 3))

results = []
with open('MicroarrayExpression.csv', encoding='us-ascii') as ma_expressions:
    #gather all results in a list
    results = [process_data(line) for line in ma_expressions]
    #this is an efficient short for
    #for line in ma_expressions:
    #    results.append(process_data(line))


#just a check because I got two probes with the same maximal expression
for probe in results:
    if probe[0] == '1058299' or probe[0] == '1070089':
        print(probe)

#print an empty line
print('\n')

#get the probes with the highest expression value and the one with the highest average
#expression value
probe_h = (0.0, 0.0, 0.0)
probe_ha = (0.0, 0.0, 0.0)
for result in results:
    #loop once to get the two probes as required. Is faster than looping twice!
    if result[1] > probe_h[1]:
        probe_h = result
    if result[2] > probe_ha[2]:
        probe_ha = result
print(f'probe: {probe_h[0]}, max expression: {probe_h[1]}, average expression {probe_h[2]}')
print(f'probe: {probe_ha[0]}, max expression: {probe_ha[1]}, average expression {probe_ha[2]}')



('1058299', 18.382, 10.01)
('1070089', 18.382, 18.113)


probe: 1058299, max expression: 18.382, average expression 10.01
probe: 1070089, max expression: 18.382, average expression 18.113


## Part 5: Integrate other sources of information

By using the other available csv files. Can you report the **gene_name** of the probe having the highest overall and highest average value? 

Expected outcome: probe: 1070089, gene name: "AGILENT probe A_23_P317056 (non-RefSeq)"

In [13]:
#Of course I could have written the code to calculate the mean myself,
#but 'a good programmer is a lazy programmer' so I use what is available
#in the standard library
from statistics import mean

def process_data(line):
    '''processes a line a specified in assignments of week 1, part 4
        input: string which is a line consisting of 1 int (the id) and many floats (the values)
        output: a tuple consisting of the id, maximum value and average value of the values
    '''
    pid, *values = line.split(',')
    for position, value in enumerate(values):
        values[position] = float(value)
    return (pid, round(max(values), 3), round(mean(values), 3))

results = []
with open('MicroarrayExpression.csv', encoding='us-ascii') as ma_expressions:
    #gather all results in a list
    results = [process_data(line) for line in ma_expressions]
    #this is an efficient short for
    #for line in ma_expressions:
    #    results.append(process_data(line))


#get the probes with the highest expression value and the one with the highest average
#expression value
probe_h = (0.0, 0.0, 0.0)
probe_ha = (0.0, 0.0, 0.0)
for result in results:
    #loop once to get the two probes as required. Is faster than looping twice!
    if result[1] > probe_h[1]:
        probe_h = result
    if result[2] > probe_ha[2]:
        probe_ha = result


with open('Probes.csv', encoding='us-ascii') as ma_probes:
    for line in ma_probes:
        #read first line (header) and ignore
        if line.startswith('probe_id'):
            pass
        else:
            probe_info = line.split(',')
            #the required probe has the same id (the first field) as probe_ha
            if probe_info[0] == probe_ha[0]:
                print(f'probe: {probe_info[0]}, gene name: {probe_info[4]}')


probe: 1070089, gene name: "AGILENT probe A_23_P317056 (non-RefSeq)"


## Part 6: Bonus (optional)

Can you report on the 5 genes that have the top 5 higest average expression value?



In [15]:
#Of course I could have written the code to calculate the mean myself,
#but 'a good programmer is a lazy programmer' so I use what is available
#in the standard library
from statistics import mean

def process_data(line):
    '''processes a line a specified in assignments of week 1, part 4
        input: string which is a line consisting of 1 int (the id) and many floats (the values)
        output: a tuple consisting of the id, maximum value and average value of the values
    '''
    pid, *values = line.split(',')
    for position, value in enumerate(values):
        values[position] = float(value)
    return (pid, round(max(values), 3), round(mean(values), 3))

results = []
with open('MicroarrayExpression.csv', encoding='us-ascii') as ma_expressions:
    #gather all results in a list
    results = [process_data(line) for line in ma_expressions]
    #this is an efficient short for
    #for line in ma_expressions:
    #    results.append(process_data(line))

#sort the list with results on the second field and return the first five
results.sort(key = lambda item: item[1], reverse = True)
top_five = results[0:5]

#some checks
#low_five = results[-5:]
#print(top_five)
#print(low_five)

#convert to dict for easier searching
top_probes = {item[0]: item[1] for item in top_five}

#put the keys of the dict in a tuple
top_probe_ids = top_probes.keys()

#get the gene name
with open('Probes.csv', encoding='us-ascii') as ma_probes:
    for line in ma_probes:
        #read first line (header) and ignore
        if line.startswith('probe_id'):
            pass
        else:
            probe_info = line.split(',')
            #the required probe has the same id (the first field) one of the top probes
            if probe_info[0] in top_probe_ids:
                print(f'probe: {probe_info[0]}, gene name: {probe_info[4]}, avg expression: {top_probes[probe_info[0]]}')



probe: 1058299, gene name: "cofilin 1 (non-muscle)", avg expression: 18.382
probe: 1056447, gene name: "glial fibrillary acidic protein", avg expression: 18.382
probe: 1055120, gene name: "cysteine-rich, avg expression: 18.382
probe: 1054004, gene name: "metallothionein 3", avg expression: 18.382
probe: 1053841, gene name: "NADH dehydrogenase (ubiquinone) 1 alpha subcomplex, avg expression: 18.382


# Conclusion

We found genes that are highly expressed in this brain dataset. Overall, detecting highly expressed genes helps researchers uncover the players that drive cellular processes or disease mechanisms. This week you took the first step in this analyis. Well done! 